### DISCUSSION TITLE
Adding a new tag in peer string `pubKeyType`

### DISCUSSION THREAD

**Discussion by colibrishin** [planetarium/libplanet#2285] [Ideas]
After (and yet) implementing multiple digital signatures, BLS in the current situation, `Address`, `Peer` could have or have been generated from any public key with different digital signatures. We might have to add a recognizable trait in the peer string before parsing the public key by changing to `<pubKeyType>,<pubKey>,<ipaddr>,<port>.`

However, This is not back-compatible, in cost, it might be too early to add a public key type in the peer string by just adding one new digital signature and not yet field-tested or clear use-cases of BLS implementation. 

We could recognize the public key type by comparison of byte length (`Secp256K1` public key length is _(130 | 66)_, `BLS` public key length is _48_.) temporarily and take advantage of compatibility.

I’d like to hear opinions of whether peer string should be affected by BLS implementation.

- Add new pubKeyType in peer string.
- Keep the peer string, and check byte representation length temporaily.
- Alternative way to recognize public key type or another reason.

**Comment by dahlia** [comment]
The indirect problem of this situation is that the current representation of `BoundPeer`[^1] is difficult to extend in the future.  It's basically a single-line <abbr title="comma-separated values">CSV</abbr> without labels.  If we need to break its backward compatibility we don't need maintain the CSV format at all — if we just add a column to the format this time we will face the same situation next time due to lack of its extensibility.  If we submit to break backward compatibility this time IMHO it'd better turn it into a totally different format, for example, [multiaddr][][^2] or [URI][][^3].

Or if we decide to stick with the current format for backward compatibility I believe our only reasonable option is to recognize key type by its length.

[^1]: FYI `Peer` was gone away recently.  See also: #2233.
[^2]: It's used by [libp2p] and IPFS.
[^3]: FYI it differs from URL.  See also *[URIs, URLs, and URNs: Clarifications and Recommendations]*.

[multiaddr]: https://multiformats.io/multiaddr/
[URI]: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier
[libp2p]: https://docs.libp2p.io/concepts/addressing/
[IPFS]: https://docs.ipfs.tech/concepts/glossary/#multiaddr
[URIs, URLs, and URNs: Clarifications and Recommendations]: https://www.w3.org/TR/uri-clarification/
